### PR TITLE
SYNR-1187

### PR DESCRIPTION
**Pull request by kimyen**

### CODE REVIEW

File: R/zzz.R [hunk 1]
```diff
@@ -123,6 +123,9 @@
     }
     lockBinding("asDataFrame", object)
   }
+  if (grepl("^GeneratorWrapper", class(object)[1])) {
+    class(object) <- "GeneratorWrapper"
```
Comment (kimyen) [inline R/zzz.R]: append this name instead of overwrite class(object)

File: R/zzz.R [hunk 2]
```diff
@@ -123,6 +123,9 @@
     }
     lockBinding("asDataFrame", object)
   }
+  if (grepl("^GeneratorWrapper", class(object)[1])) {
```
Comment (brucehoff) [inline R/zzz.R]: .modify -> ".objectDefinitionHelper"

File: inst/python/gateway.py [hunk 3]
```diff
@@ -6,16 +6,23 @@
...
+      if self._use_list:
+        raise "asList() has been used."
+      self._use_iter = True
+      return self._inner.__next__()
+
+    def asList(self):
+      if self._use_iter:
+        raise "nextElem() has been used."
+      if self._use_list:
+        raise "asList() can be used only once."
```
Comment (brucehoff) [inline inst/python/gateway.py]: the error messages, which will be shown in R, mention Python method names.

File: inst/python/gateway.py [hunk 4]
```diff
@@ -6,16 +6,23 @@
...
-    def __getattr__(self, attr):
-        return getattr(self.inner, attr)
-
-    def __setattr__(self, attr, value):
-        setattr(self.inner, attr, value)
-    
     def nextElem(self):
-        return self.__next__()
+      if self._use_list:
+        raise "asList() has been used."
```
Comment (brucehoff) [inline inst/python/gateway.py]: "Have already enumerated all elements."
(Or just remove this check and let the underlying generator raise an error.)
The latter should only be done if the message is informative.

Comment (kimyen) [reply inst/python/gateway.py]: The original error: 
```
Error in pyCall("__R__.namespace[2].nextElem", args = xargs, kwargs = xkwargs) : 
  error occurred while calling '__R__.namespace[2].nextElem'
```

**Pull request merged**